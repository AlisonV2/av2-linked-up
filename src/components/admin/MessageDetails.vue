<template>
  <div id="chat-window">
    <div class="col-12 mb-4" v-for="message in messages" :key="message.sender">
      <div
        class="chat-container mb-0"
        :class="[userId == message.sender ? 'darker' : '']"
      >
        <div :class="[userId == message.sender ? 'text-end' : '']">
          <span class="name">{{ message.senderName }}</span>
          <p class="message">{{ message.message }}</p>
        </div>
      </div>
      <div :class="[userId == message.sender ? 'text-end' : '']">
        <small class="text-muted px-3">{{ message.date }}</small>
      </div>
    </div>
  </div>
  <div class="row chat-row">
    <form @submit.prevent="sendMessage">
      <div class="input-group mb-3">
        <input
          class="form-control"
          type="textarea"
          placeholder="Your message"
          v-model="newMessage"
        />
        <button class="btn btn-outline-secondary" type="submit">Send</button>
      </div>
    </form>
  </div>
</template>

<script>
import * as Sentry from '@sentry/vue';

/**
 * @exports MessageDetails
 * @type {Page}
 * @requires Sentry
 * @vue-data {array} messages
 * @vue-data {string} newMessage
 * @vue-data {string} userId
 */
export default {
  name: 'MessageDetails',
  data() {
    return {
      messages: [],
      userId: '',
      newMessage: '',
    };
  },
  created() {
    this.$store
      .dispatch('getCurrentUser')
      .then(() => {
        this.userId = this.$store.getters.getCurrentUser;
      })
      .catch((err) => {
        Sentry.captureException(err);
      });

    this.getMessages();
  },
  methods: {
    sendMessage() {
      const projectId = this.$route.params.id;
      const message = {
        projectId: projectId,
        message: this.newMessage,
      };
      this.$store
        .dispatch('sendMessage', message)
        .then(() => {
          this.getMessages();
        })
        .then(() => {
          this.newMessage = '';
        })
        .catch((err) => {
          Sentry.captureException(err);
        });
    },
    getMessages() {
      this.$store
        .dispatch('getMessages', this.$route.params.id)
        .then(() => {
          this.messages = this.$store.getters.getMessages;
        })
        .then(() => {
          const element = document.querySelector('#chat-window');
          element.scrollTop = element.scrollHeight;
        })
        .catch((err) => {
          Sentry.captureException(err);
        });
    },
  },
};
</script>

<style lang="scss" scoped>
#chat-window {
  overflow-y: scroll;
  height: 70vh;
}
.name {
  font-weight: bold;
  margin-right: 6px;
}

.message {
  margin-bottom: 0;
}
/* Chat containers */
.chat-container {
  border: 2px solid #dedede;
  background-color: #f1f1f1;
  border-radius: 5px;
  padding: 10px;
  margin: 10px 0;
}

/* Darker chat container */
.darker {
  border-color: #ccc;
  background-color: #ddd;
}

/* Clear floats */
.container::after {
  content: '';
  clear: both;
  display: table;
}

/* Style images */
.container img {
  float: left;
  max-width: 60px;
  width: 100%;
  margin-right: 20px;
  border-radius: 50%;
}

/* Style the right image */
.container img.right {
  float: right;
  margin-left: 20px;
  margin-right: 0;
}

/* Style time text */
.time-right {
  float: right;
  color: #aaa;
}

/* Style time text */
.time-left {
  float: left;
  color: #999;
}

.chat-row {
  position: absolute;
  bottom: 2rem;
  width: 70%;
}
</style>
