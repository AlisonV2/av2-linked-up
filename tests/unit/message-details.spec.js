import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import MessageDetails from '@/components/admin/MessageDetails.vue';

const actions = {
  getMessages: jest.fn(),
  getCurrentUser: jest.fn(),
  sendMessage: jest.fn(),
};

const getters = {
  getMessages: jest.fn(),
  getCurrentUser: jest.fn(),
};

const store = createStore({
  actions,
  getters,
});

const $route = {
  params: { id: 1 },
};

const config = {
  data() {
    return {
      messages: [],
      userId: '',
      newMessage: ''
    };
  },
  global: {
    plugins: [store],
    mocks: { $route },
  },
};

/**
 * @module MessageDetailsTest
 */
describe('MessageDetails.vue', () => {
  it('MessageDetails Snapshot', () => {
    const component = shallowMount(MessageDetails, config);
    expect(component.element).toMatchSnapshot();
  });

  it('getMessages action should have been called', async () => {
    const component = shallowMount(MessageDetails, config);
    MessageDetails.created.call(component.vm);
    await component.vm.$nextTick();
    expect(actions.getMessages).toHaveBeenCalled;
  });

  it('getMessages getter should have been called', async () => {
    const component = shallowMount(MessageDetails, config);
    MessageDetails.created.call(component.vm);
    await component.vm.$nextTick();
    expect(getters.getMessages).toHaveBeenCalled;
  });
  it('Dispatches sendMessage action on submit', async () => {
    const component = shallowMount(MessageDetails, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.sendMessage).toHaveBeenCalled();
  });
  it('getMessages action should be called on submit', async () => {
    const component = shallowMount(MessageDetails, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.getMessages).toHaveBeenCalled();
  });
});
