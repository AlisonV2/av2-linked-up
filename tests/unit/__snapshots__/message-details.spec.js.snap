// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageDetails.vue MessageDetails Snapshot 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="row chat-window"
    id="chatWindow"
  >
    
    
  </div>
  <div
    class="row chat-row"
  >
    <form>
      <div
        class="input-group mb-3"
      >
        <input
          class="form-control"
          placeholder="Your message"
          type="textarea"
        />
        <button
          class="btn btn-outline-secondary"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </div>
  
</div>
`;
